tests/auto/ci_auto.py

```diff
@@ -211,18 +216,21 @@ def job_failed(self, logger, job_name, exception=Exception, STDOUT=False,
             logger.critical(f'STDOUT: {[item for item in out if not None]}')
             logger.critical(f'STDERR: {[eitem for eitem in err if not None]}')
 
-            
+
 def setup_env():
     logger = logging.getLogger('SETUP')
 
     hostname = os.getenv('HOSTNAME')
     if bool(re.match(re.compile('hfe.+'), hostname)):
         machine = 'hera'
+        hpc_acc = 'zrtrr'
     elif bool(re.match(re.compile('hecflow.+'), hostname)):
         machine = 'hera'
+        hpc_acc = 'zrtrr'
     elif bool(re.match(re.compile('fe.+'), hostname)):
         machine = 'jet'
-        os.environ['ACCNR'] = 'h-nems'
+        hpc_acc = 'nrtrr'
+        os.environ['ACCNR'] = hpc_acc
     elif bool(re.match(re.compile('gaea.+'), hostname)):
         machine = 'gaea'
         os.environ['ACCNR'] = 'nggps_emc'
```

Comment:
Is there a way to avoid the code "knowing" which machine it is on?  Perhaps by passing this information in with a configuration?  Using a config to specify that would allow for a completely general solution that will work anywhere without modifying source code.

Reply:
What if I passed the machine in as a command line argument, and then had one config file with an entry for each machine specifying the account and work directory? I'm not really sure about the lines that put out the account code as an environment variable - that was in the model testing code. I'd like to avoid creating a different config file for each machine, though I guess that's how the environment files are done in ufs-srweather-app. Both code and config files are in a repository, so to me it seems the same amount of work to update either way - more work for multiple files.

Reply:
That would certainly be better.  I would support that.  Ideally there would be a single config file that is passed as an argument that contains all the configuration information the code needs.  That would include (but not limited to) one entry for the machine, one entry for the account, and one entry for the work directory.  To use the tool you have to "configure" it by setting the options in the config file.  That would avoid storing machine names and sensitive information like project accounts and paths in a public repo.  It would also make it completely general such that you could configure lots of different types of situations by modifying a single file and without touching any code.

Reply:
I was thinking of one config file, and use command line arg of machine to choose:
[hera]
hpc_acc=zrtrr
workdir=/scratch2/BMC/zrtrr/rrfs_ci/autoci/pr

[jet]
hpc_acc=nrtrr
workdir=/lfs1/BMC/nrtrr/rrfs_ci/autoci/pr

But it sounds like you're suggesting:
on hera have a config file (that's not in github):
machine=hera
hpc_acc=zrtrr
workdir=/scratch2/BMC/zrtrr/rrfs_ci/autoci/pr

on jet have a config file (that's not in github):
machine=jet
hpc_acc=nrtrr
workdir=/lfs1/BMC/nrtrr/rrfs_ci/autoci/pr

Am I understanding correctly?

Reply:
I think you could stage the same file will all the info about our supported machines. Then you'd just pass the machine string and the path to the staged file. Maybe hpc_acc should just be a command line option if we're worried about storing it in a repo. That config file should def be stored with the software, IMO.

Reply:
Yes.  That is how I interpreted your suggestion.  I was taking it a step further and passing the config instead of the machine where the config would just be:

```
machine=some_machine
hpc_acc=project_account
workdir=/path/to/autoci/pr
```

And that is it.  One config file on github that has placeholders or default values.  No stanzas for particular machines.  If the testing infrastructure didn't need to know the machine name, you wouldn't even need to specify that either.  Then, users (probably us) just modify the config to use the appropriate values for their situation where they've checked out the repo and are using it.

Reply:
If you wanted to, you could maintain a config file in the repo for each machine.  Like is done for the UFS with the `build_*.env` files.  But then you have to maintain multiple things.  This way you don't have to maintain ANY platform specific information at all.  Which means it will never be out of date or incorrect.

Reply:
My preference would be for the words "jet", "hera", "orion", etc to only appear in documentation and no where else.  Not in the default config file stored in the repo.  And definitely not in the code.  If setting the options is for some reason complex or not straight forward, then that problem can be solved by the documentation in the README.md file or wiki.

Reply:
@christopherwharrop-noaa How would we manage that knowledge for re-deploying, or backing up configurations?

We should keep config files under version control, no?

Reply:
You don't need to keep the exact config file you are using for a particular deployment in a repo.  All you need is the default config file.  The knowledge, if there is something special, can be in the documentation, which would also be the backup.  The repo should store the software and a default configuration, not how it is configured for a particular deployment for a particular scenario.

Reply:
Ok, this is getting into a bit of a grey area.  If you want to use @venitahagerty 's proposed solution with a single config file with all the machine stuff in it I'm ok with it.  What I am not ok with is having machine names appear anywhere in the code.  I can understand a desire to track changes to configs for a particular deployment so you have a history of it.  Where I get hung up is when you put those into the same repository for the software you are configuring.  I'm not sure it belongs there because it muddies the waters between development of the software, and specific deployments of the software.  Those two objectives have nothing to do with each other.  One is development of the software itself.  The other is tracking how the software is configured for a particular purpose which is completely outside the scope of the software development.  To be general and keep things separate, details of how the software is used in particular deployments should not be kept with the software itself. IMHO.

Reply:
I've spent a lot of time re-reading these comments and thinking about the pros and cons of the different methods.
 
I'm thinking:
- have one config file stored in the repo with default values and a fixed name
- user will have responsibility to update the config file on each HPC
- we could also put the name of the app in the config file, for future expansion of testing
- once this was done, a machine name would only appear in code as a comment of an example tag
- document what user should do
- contents of default config file CImachine.cfg:
machine=some_first_tier_machine
hpc_acc=project_account
weather_app=some_ufs_weather_app
workdir=/path/to/autoci/pr

From a re-running the test perspective, once the config file is set up on an HPC machine, it probably won't need to be changed very often. The code manager will apply a label (i.e. ci-hera-intel-WE) to a pull request, and the test will run.

For the convenience of the code manager and to reduce costs, the code manager decides which of the first tier HPC machines should run which test by the label that they apply. This is similar to how model testing is done.

I wrestled with machine and/or account as command line args, and with this method it seemed better and more consistent to put all the values in a config file. Like Chris suggested above, and similar to how config.sh is edited for the workflow experiment.

Does this work for everybody?

Reply:
@venitahagerty - I am happy with your latest proposal.  It addresses my concerns completely.  If anyone would like an expanded explanation of my suggestion please let me know.

Reply:
Sorry to have been absent from the conversation. @venitahagerty, this is a lot of great work. My fluency in Python is limited, so I don't have much to suggest in terms of the syntax/implementation.

Your outlined strategy above is more modular than the ```if-else``` block and will also be easier to maintain. Abstracting the machine and associated config files the source code runs on/uses will make future development easier. (Notably for me in making modifications to include using different architectures in the short term).

Reply:
Looking at code again, I had found a different way (based on the repo config file) to solve the problem of the name of the weather app, so I won't include the name of the app in the config file.

Reply:
I think the proposal looks pretty good, and seems to address most of the concerns. I may not be seeing one part clearly though.

How do we test to insure that comments are consistent with requirements of what is provided in the application, and that our staged config file meets those requirements?

Reply:
When you say that comments are consistent with requirements, I'm not sure what you mean.

Reply:
The proposal suggests that we are documenting the requirements of the config file (not holding a config file in the repo). I'm considering a situation where the requirements change. We have to update the documentation, the code, and the config file to reflect the same information. 

An example...if we deployed the system with the variables you suggest above on several machines. Then we come back later, we will need to change that file on each machine individually, and manually ensure that it is consistent with the new requirements BEFORE we deploy, and make sure that we also updated documentation for what to include in the new files.

Reply:
We would be keeping a default/sample config file (CImachine.cfg) in the repo. A config file is for only one machine. It would need to have values inserted in each copy on an HPC machine. If we needed to add a new variable, we would have to change the config file in the repo and the code, and on each machine, and also update the documentation. I don't see a way to avoid that.

However, potentially sensitive information is kept out of the repo. The machine name and account number are not in the code. A default/model config file is in the repo as an example. And to add a new machine, you clone the repo on the new machine and update the config file - no code changes.

Reply:
I'm _not_ asking you to do this here, but maybe we could have some code that will do that deployment for us. If we have the "tested" sample config be the basis for updating the documentation, and also tell us how to fix/update a deployed machine config, it could be helpful. Just a (probably not fully coherent) thought.

I'd say go for the proposal!

Reply:
I have pushed the new code, after successful testing on hera.